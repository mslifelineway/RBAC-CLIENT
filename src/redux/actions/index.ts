export * from "./auth.action";
export * from "./account.action";
