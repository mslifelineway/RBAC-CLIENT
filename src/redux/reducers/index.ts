import authReducer from "./auth.slice";
import accountReducer from "./account.slice";
import roleReducer from "./role.slice";
import permissionReducer from "./permission.slice";

export const reducers = {
  authReducer,
  accountReducer,
  roleReducer,
  permissionReducer,
};
