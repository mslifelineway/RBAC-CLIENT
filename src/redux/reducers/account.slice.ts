import { createSlice } from "@reduxjs/toolkit";
import { createAccount } from "../actions";

interface InitialState {
  loading: boolean;
  error: string | null;
}

const initialState: InitialState = {
  loading: false,
  error: null,
};

const accountSlice = createSlice({
  name: "accountSlice",
  initialState,
  reducers: {},
  extraReducers: (builder) => {
    builder
      .addCase(createAccount.pending, (state) => {
        state.loading = true;
        state.error = null;
      })
      .addCase(createAccount.fulfilled, (state, action) => {
        state.loading = false;
      })
      .addCase(createAccount.rejected, (state, action) => {
        state.loading = false;
        state.error = action.payload as string;
      });
  },
});

export const {} = accountSlice.actions;
export default accountSlice.reducer;
