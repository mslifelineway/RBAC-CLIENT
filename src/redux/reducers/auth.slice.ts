import { createSlice } from "@reduxjs/toolkit";
import { login, logout } from "../actions";

export interface IPermission {
  _id: string;
  name: string;
  permissionUniqueKey: string;
  description: string;
}

export interface IRole {
  _id: string;
  name: string;
  description: string;
  permissions: IPermission[];
}

export type IAdministrator = {
  _id: string;
  email: string;
  firstName: string;
  lastName: string;
  phoneNumber: string;
  roles: string[] | IRole[];
  permissionUniqueKeys: string[];
};

export type IEmployee = {
  _id: string;
  email: string;
  firstName: string;
  lastName: string;
  phoneNumber: string;
  roles: string[];
  permissionUniqueKeys: string[];
};

export interface InitialState {
  isAuthenticating: boolean;
  isAuthenticated: boolean;
  error: string | null;
  data: IAdministrator | IEmployee | null;
}

const initialState: InitialState = {
  isAuthenticating: false,
  isAuthenticated: false,
  error: null,
  data: null,
};

const authSlice = createSlice({
  name: "authSlice",
  initialState,
  reducers: {
    setLoginTestData(state, action) {
      state.data = action.payload;
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(login.pending, (state) => {
        state.isAuthenticating = true;
        state.error = null;
      })
      .addCase(login.fulfilled, (state, action) => {
        state.isAuthenticating = false;
        state.isAuthenticated = true;
        state.data = action.payload;
      })
      .addCase(login.rejected, (state, action) => {
        state.isAuthenticating = false;
        state.error = action.payload as string;
      })
      .addCase(logout.pending, (state) => {
        state.isAuthenticating = true;
        state.error = null;
      })
      .addCase(logout.fulfilled, (state) => {
        state.isAuthenticating = false;
        state.isAuthenticated = false;
        state.data = null;
      })
      .addCase(logout.rejected, (state, action) => {
        state.isAuthenticating = false;
        state.error = action.payload as string;
      });
  },
});

export const { setLoginTestData } = authSlice.actions;
export default authSlice.reducer;
