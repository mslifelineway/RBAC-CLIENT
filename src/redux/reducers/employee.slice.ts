import { createSlice } from "@reduxjs/toolkit";
import { assignRoles, createEmployee, fetchEmployees } from "../actions";
import { IEmployee, IRole } from "../../types";

interface InitialState {
  loading: boolean;
  isCreating: boolean;
  isRoleAssigning: boolean;
  error: string | null;
  data: IEmployee[];
}

const initialState: InitialState = {
  loading: false,
  isCreating: false,
  isRoleAssigning: false,
  error: null,
  data: [],
};

const employeeSlice = createSlice({
  name: "employeeSlice",
  initialState,
  reducers: {},
  extraReducers: (builder) => {
    builder
      .addCase(fetchEmployees.pending, (state) => {
        state.loading = true;
        state.error = null;
      })
      .addCase(fetchEmployees.fulfilled, (state, action) => {
        state.loading = false;
        state.data = action.payload;
      })
      .addCase(fetchEmployees.rejected, (state, action) => {
        state.loading = false;
        state.error = action.payload as string;
      })
      .addCase(createEmployee.pending, (state) => {
        state.isCreating = true;
        state.error = null;
      })
      .addCase(createEmployee.fulfilled, (state, action) => {
        state.isCreating = false;
        state.data.push(action.payload);
      })
      .addCase(createEmployee.rejected, (state, action) => {
        state.isCreating = false;
        state.error = action.payload as string;
      })
      .addCase(assignRoles.pending, (state) => {
        state.isRoleAssigning = true;
        state.error = null;
      })
      .addCase(assignRoles.fulfilled, (state, action) => {
        state.isRoleAssigning = false;
        const data = state.data.find((e) => e._id === action.payload._id);
        if (data) {
          data.roles = action.payload.roles;
        }
      })
      .addCase(assignRoles.rejected, (state, action) => {
        state.isRoleAssigning = false;
        state.error = action.payload as string;
      });
  },
});

export const {} = employeeSlice.actions;
export default employeeSlice.reducer;
