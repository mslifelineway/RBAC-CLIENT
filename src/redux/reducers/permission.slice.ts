import { createSlice } from "@reduxjs/toolkit";
import { fetchPermissions } from "../actions";
import { IPermission } from "../../types";

interface InitialState {
  loading: boolean;
  error: string | null;
  data: IPermission[];
}

const initialState: InitialState = {
  loading: false,
  error: null,
  data: [],
};

const permissionSlice = createSlice({
  name: "permissionSlice",
  initialState,
  reducers: {},
  extraReducers: (builder) => {
    builder
      .addCase(fetchPermissions.pending, (state) => {
        state.loading = true;
        state.error = null;
      })
      .addCase(fetchPermissions.fulfilled, (state, action) => {
        state.loading = false;
        state.data = action.payload;
      })
      .addCase(fetchPermissions.rejected, (state, action) => {
        state.loading = false;
        state.error = action.payload as string;
      });
  },
});

export const {} = permissionSlice.actions;
export default permissionSlice.reducer;
