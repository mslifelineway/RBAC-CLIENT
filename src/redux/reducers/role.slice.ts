import { createSlice } from "@reduxjs/toolkit";
import { fetchRoles } from "../actions";
import { IRole } from "../../types";

interface InitialState {
  loading: boolean;
  error: string | null;
  data: IRole[];
}

const initialState: InitialState = {
  loading: false,
  error: null,
  data: [],
};

const roleSlice = createSlice({
  name: "roleSlice",
  initialState,
  reducers: {},
  extraReducers: (builder) => {
    builder
      .addCase(fetchRoles.pending, (state) => {
        state.loading = true;
        state.error = null;
      })
      .addCase(fetchRoles.fulfilled, (state, action) => {
        state.loading = false;
        state.data = action.payload;
      })
      .addCase(fetchRoles.rejected, (state, action) => {
        state.loading = false;
        state.error = action.payload as string;
      });
  },
});

export const {} = roleSlice.actions;
export default roleSlice.reducer;
