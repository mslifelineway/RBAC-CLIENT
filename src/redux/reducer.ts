import { combineReducers } from "@reduxjs/toolkit";
import { reducers } from "./reducers";

export const rootReducer = combineReducers(reducers);
