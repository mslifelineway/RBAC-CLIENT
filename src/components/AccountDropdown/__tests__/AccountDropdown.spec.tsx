import { act, create } from "react-test-renderer";
import { AccountDropdown } from "../";
import { Provider } from "react-redux";
import { configureStore } from "@reduxjs/toolkit";
import { reducers } from "../../../redux/reducers";
import {
  IAdministrator,
  setLoginTestData,
} from "../../../redux/reducers/auth.slice";

jest.mock("react-dom", () => {
  const original = jest.requireActual("react-dom");

  return {
    ...original,
    createPortal: (node: JSX.Element) => node,
  };
});

const mockStore = configureStore({
  reducer: {
    authReducer: reducers.authReducer,
  },
});

describe("AccountDropdown Component", () => {
  it("renders correctly with default props", () => {
    const component = create(
      <Provider store={mockStore}>
        <AccountDropdown />
      </Provider>
    );
    expect(component.toJSON).toMatchSnapshot();
  });
  it("renders roles of current user in account dropdown ", () => {
    const initialState: IAdministrator = {
      _id: "661ab9f1c3993d0eb721e911",
      email: "admin@gmail.com",
      firstName: "test",
      lastName: "me",
      phoneNumber: "7015720216",
      roles: ["ADMINISTRATOR"],
      permissions: [],
    };
    act(() => {
      mockStore.dispatch(setLoginTestData(initialState));
    });

    const component = create(
      <Provider store={mockStore}>
        <AccountDropdown />
      </Provider>
    );
    const rolesElement = component.root.findByProps({ "data-testid": "roles" });

    expect(component.toJSON()).toMatchSnapshot();
    expect(rolesElement.props.children).toEqual(
      `(${initialState.roles.toString()})`
    );
  });
});
