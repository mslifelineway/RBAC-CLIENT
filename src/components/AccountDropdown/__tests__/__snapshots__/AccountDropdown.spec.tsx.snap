// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountDropdown Component renders correctly with default props 1`] = `[Function]`;

exports[`AccountDropdown Component renders roles of current user in account dropdown  1`] = `
Array [
  <button
    aria-controls=":r7:"
    aria-expanded={false}
    aria-haspopup="menu"
    className="base-MenuButton-root css-saewb7"
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    tabIndex={0}
    type="button"
  >
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge css-tzssek-MuiSvgIcon-root"
      data-testid="AccountCircleOutlinedIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2M7.35 18.5C8.66 17.56 10.26 17 12 17s3.34.56 4.65 1.5c-1.31.94-2.91 1.5-4.65 1.5s-3.34-.56-4.65-1.5m10.79-1.38C16.45 15.8 14.32 15 12 15s-4.45.8-6.14 2.12C4.7 15.73 4 13.95 4 12c0-4.42 3.58-8 8-8s8 3.58 8 8c0 1.95-.7 3.73-1.86 5.12"
      />
      <path
        d="M12 6c-1.93 0-3.5 1.57-3.5 3.5S10.07 13 12 13s3.5-1.57 3.5-3.5S13.93 6 12 6m0 5c-.83 0-1.5-.67-1.5-1.5S11.17 8 12 8s1.5.67 1.5 1.5S12.83 11 12 11"
      />
    </svg>
  </button>,
  <div
    className="base-Popup-root base-Menu-root"
    style={
      Object {
        "left": 0,
        "position": "absolute",
        "top": 0,
        "visibility": "hidden",
      }
    }
  >
    <div
      className="placement-bottom closed"
      onTransitionEnd={[Function]}
    >
      <ul
        className="base-Menu-listbox css-grlzle"
        id=":r7:"
        onBlur={[Function]}
        onKeyDown={[Function]}
        role="menu"
        tabIndex={-1}
      >
        <p
          data-testid="roles"
          style={
            Object {
              "fontStyle": "italic",
              "textAlign": "center",
            }
          }
        >
          (ADMINISTRATOR)
        </p>
        <li
          className="base-MenuItem-root css-v5whyu"
          id=":r8:"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onPointerOver={[Function]}
          role="menuitem"
          tabIndex={-1}
        >
          Profile
        </li>
        <li
          className="base-MenuItem-root css-v5whyu"
          id=":ra:"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onPointerOver={[Function]}
          role="menuitem"
          tabIndex={-1}
        >
          Language settings
        </li>
        <li
          className="base-MenuItem-root css-v5whyu"
          id=":rc:"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onPointerOver={[Function]}
          role="menuitem"
          tabIndex={-1}
        >
          Log out
        </li>
      </ul>
    </div>
  </div>,
]
`;
